### ISSUE TITLE
Minimization of DFA from finite language makes the language infinite

### ISSUE THREAD

**Issue by martinec** [caleb531/automata#182]
While using the `DFA.from_finite_language()` method to create a basic DFA from a finite language and then applying dfa.minify(), the resulting DFA incorrectly introduces a loop that allows an infinite number of accepted words. This behavior deviates from the expected result where the minimized DFA should recognize the same finite language as the unminimized DFA.

**Steps to Reproduce:**

1. Create a DFA from a finite language using DFA.from_finite_language() with the language {'ab', 'abcb'} and input symbols {'a', 'b', 'c'}.
1. Print the maximum word length and the DFA representation.
1. Apply the minify() method on the DFA.
1. Print the maximum word length and the minimized DFA representation.

**Code**

```python
dfa = DFA.from_finite_language(
    language={'ab', 'abcb'},
    input_symbols={'a', 'b', 'c'}
)
print("unminified maximum_word_length:", dfa.maximum_word_length())
print("unminified:", dfa)
dfa = dfa.minify()
print("minified maximum_word_length:", dfa.maximum_word_length())
print("minified:", dfa)
```

**Output**
```shell
unminified maximum_word_length: 4
unminified: DFA(states={'', 'ab', 'abc', 'abcb', 'a'}, input_symbols={'a', 'c', 'b'}, transitions={'': {'a': 'a'}, 'a': {'b': 'ab'}, 'ab': {'c': 'abc'}, 'abc': {'b': 'abcb'}, 'abcb': {}}, initial_state='', final_states={'ab', 'abcb'}, allow_partial=True)
minified maximum_word_length: None
minified: DFA(states={0, 1, 2}, input_symbols={'a', 'c', 'b'}, transitions={0: {'a': 2}, 1: {'c': 2}, 2: {'b': 1}}, initial_state=0, final_states={1}, allow_partial=True)
DFA(states={0, 1, 2, 3}, input_symbols={'a', 'c', '<E>', 'b'}, transitions={0: {'a': 2}, 1: {'c': 2, '<E>': 3}, 2: {'b': 1}, 3: {}}, initial_state=0, final_states={3}, allow_partial=True)
```

**DFA**

<img src="https://github.com/caleb531/automata/assets/800627/e16afef2-d87c-4d00-ba77-0f15e578e5c9" width="400">

**Minified DFA**

<img src="https://github.com/caleb531/automata/assets/800627/625b4067-b190-4f60-9f62-2fa46602ddd3" width="200">

**Environment**:

- Python version: 3.9
- automata-lib version: 8.1.0
- Operating System: Debian

The problem seems to be related to the minimization algorithm, potentially in the way it handles transitions and final states during state reduction, but I may have missed something, thanks for your help !

**Comment by eliotwrobson** [comment]
@martinec thank you for reporting this! Can you try this again while setting `allow_partial=False` in the constructor for the first DFA? This might be related to weirdness with minimizing partial DFAs, as support for that was only recently expanded.

**Comment by martinec** [comment]
@eliotwrobson Thanks for your help.

Using the following code:

```python
    dfa = DFA.from_finite_language(
    language={'ab', 'abcb'},
    input_symbols={'a', 'b', 'c'}, as_partial=False)
    print("unminified maximum_word_length:", dfa.maximum_word_length())
    print("unminified:", dfa)
    dfa = dfa.minify()
    print("minified maximum_word_length:", dfa.maximum_word_length())
    print("minified:", dfa)
    dfa = dfa.to_partial()
    print("partial minified maximum_word_length:", dfa.maximum_word_length())
    print("partial minified:", dfa)
```

The output is:

```python
unminified maximum_word_length: 4
unminified: DFA(states={'', 0, 'ab', 'abcb', 'abc', 'a'}, input_symbols={'b', 'c', 'a'}, transitions={'': {'b': 0, 'c': 0, 'a': 'a'}, 'a': {'b': 'ab', 'c': 0, 'a': 0}, 'ab': {'b': 0, 'c': 'abc', 'a': 0}, 'abc': {'b': 'abcb', 'c': 0, 'a': 0}, 'abcb': {'b': 0, 'c': 0, 'a': 0}, 0: {'b': 0, 'c': 0, 'a': 0}}, initial_state='', final_states={'ab', 'abcb'}, allow_partial=False)
minified maximum_word_length: 4
minified: DFA(states={0, 1, 2, 3, 4, 5}, input_symbols={'b', 'c', 'a'}, transitions={0: {'b': 0, 'c': 0, 'a': 0}, 1: {'b': 0, 'c': 0, 'a': 0}, 2: {'b': 1, 'c': 0, 'a': 0}, 3: {'b': 0, 'c': 2, 'a': 0}, 4: {'b': 0, 'c': 0, 'a': 5}, 5: {'b': 3, 'c': 0, 'a': 0}}, initial_state=4, final_states={1, 3}, allow_partial=False)
partial minified maximum_word_length: 2
partial minified: DFA(states={0, 1, 2}, input_symbols={'b', 'c', 'a'}, transitions={0: {'a': 2}, 1: {}, 2: {'b': 1}}, initial_state=0, final_states={1}, allow_partial=True)
```

Now, when calling dfa.to_partial(), it should give a deep copy of the dfa that is equivalent in the sense that it recognizes the same language, but you can notice the language of the DFA has changed after this operation. Any insight ?

**Comment by eliotwrobson** [comment]
I think this might be an issue with the function that gets the length of the longest word accepted. I'll play around with this more later and get back to you.

**Comment by martinec** [comment]
For reference, can you confirm that in the to_partial() function:

- [live_states](https://github.com/caleb531/automata/blob/44503d7473dc0d84e82f4f0897137d0cbf5d55e0/automata/fa/dfa.py#L237), computed as:

```python
        live_states = nx.descendants(graph, self.initial_state) | {self.initial_state}
```

The live_states are the direct descendants of the initial state and not any state from which a final state is reachable ?

- [non_trap_states](https://github.com/caleb531/automata/blob/44503d7473dc0d84e82f4f0897137d0cbf5d55e0/automata/fa/dfa.py#L238), computed as:

```python
 non_trap_states = set(self.final_states).union(
            *(nx.ancestors(graph, state) for state in self.final_states)
        )
```
The non_trap_states are only the ancestors of final states and do not includes states that are part of a path to a final state, even if they are not ancestors ?

It seems that the code might be mistakenly dropping states that, while not directly connected to the final states, are still necessary for creating longer words that the DFA should accept.

Thanks again for your help !

**Comment by eliotwrobson** [comment]
@martinec Ahh, I think that might be the culprit. I'll have some time this evening so I can try to get a fix for this out.

**Comment by eliotwrobson** [comment]
@martinec #184 fixing this bug has been merged to develop. We will try and make a new release soon, but in the meantime if you need fixed code, that should be available from that branch.

**Comment by martinec** [comment]
@eliotwrobson Understood ! thanks for your commitment

**Comment by eliotwrobson** [comment]
Fix included in the 8.2 release, closing this issue.
